Added /raw endpoint which gives an easy access to raw value of sensor

mobrembski opened the pull request:
This will give you ability to get easy-parseable values for small embedded devices results
using only simple busybox tools such as awk or sed.

mobrembski committed:
Added /raw endpoint which gives an easy access to raw value of sensor

This will give you ability to get easy-parseable values for small embedded devices results
using only simple busybox tools such as awk or sed.

uzi18 commented:
I like it, maybe you can add also option to get all vars in task, it will be far more economic.
@TD-er what do you think?

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,41 @@
+    {
+      LoadTaskSettings(taskNr);
+      htmlData += String(ExtraTaskSettings.TaskDeviceValueNames[valNr]);
+      htmlData += " ";
```
Just as a general hint. When adding a single character, it takes less resources when appending a single char instead of a single character string.
```c++
htmlData += ' ';
```

mobrembski replied:
Ahh, Right.
I will fix that, and add option to get all vars in task :)
Will be back to you guys soon :)

TD-er replied:
Maybe get simple CSV formatted output?
Or do you just separate them using a newline?
And not entirely sure if we enforce not using spaces in value names.

mobrembski replied:
Good idea with this semicolons.
I think semicolon separating value with name will be better than spaces.
And newline as a separator between variables.
I will change it.

TD-er replied:
So 1 line with only the names and a second one with only the values?
Maybe also have an extra GET parameter then to leave away the names to reduce the resource usage even more.

mobrembski replied:
I planned sth like this:
If user sends only tasknr, ex. http://<ip>/raw?tasknr=1 then it will send all variables for selected task
Temperature;23.5
Humidity;80
etc.

It could be possible as uzi18 suggested, to filter only one value, ex:
http://<ip>/raw?tasknr=1&valnr=1
Temperature;23.5

;)

TD-er commented:
> I like it, maybe you can add also option to get all vars in task, it will be far more economic.
> @TD-er what do you think?

There is already a JSON endpoint to get those.
`http://<ip>/json?tasknr=1`
I just tested it and it returns invalid JSON, so let me at least fix that bug.

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,41 @@
+    htmlData += F("ERROR: ValueId not valid!\n");
+  }
 
+  TXBuffer.startJsonStream();
```
Why start a JSON stream here and not output JSON formatted values?

mobrembski replied:
Good question, as i'm new to ESPEasy...Should i invoke some start()/end() function before sending plain HTTP text?

TD-er replied:
Well, if you need a different mime type then we may need to add a different startStream function.
So I was just wondering if we now violate some standards here.

mobrembski replied:
It need to stay as it is, because without it browser won't recognize newline characters in plain text.

TD-er replied:
That's OK.
Like I said, I just wondered whether we're violating some standards now.

mobrembski commented:
> > I like it, maybe you can add also option to get all vars in task, it will be far more economic.
> > @TD-er what do you think?
> 
> There is already a JSON endpoint to get those.
> `http://<ip>/json?tasknr=1`
> I just tested it and it returns invalid JSON, so let me at least fix that bug.

Yes, but it still get you in JSON format. Which is not always desired format. My idea was to get endpoint so we can get easy VariableName VariableValue strings, without packing in any JSON.

mobrembski committed:
Changed format of RAW to CSV, added ability to filter values

mobrembski commented:
I've updated pull request.
Added option to filter values or see only selected one, and changed format to CSV.
 TXBuffer.startJsonStream(); need to stay as it is, because without it browser won't recognize newline characters.

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -16,10 +16,11 @@ void handle_rawval()
-  const bool valueNumberValid = validTaskVarIndex(valNr);
-  if (!valueNumberValid)
+  bool valueNumberValid = true;
+  if (valNr != INVALID_TASKVAR_INDEX && !validTaskVarIndex(valNr))
```
See the code where it is checked:
```c++
taskVarIndex_t INVALID_TASKVAR_INDEX = VARS_PER_TASK;

bool validTaskVarIndex(taskVarIndex_t index) {
  return index < VARS_PER_TASK;
}
```
So this check may fail when `valNr == INVALID_TASKVAR_INDEX`

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -30,9 +31,17 @@ void handle_rawval()
+      {
+        if (valNr == INVALID_TASKVAR_INDEX || valNr == x)
+        {
+          htmlData += String(ExtraTaskSettings.TaskDeviceValueNames[x]);
```
Please try to make some estimate on the expected size of `htmlData` and call `htmlData.reserve()` before appending.
For example use an estimate of 16 bytes per var name and 6 per value + ';' + '\n' = 24 bytes per value.
So a rough estimate can be to reserve 128 bytes..

TD-er commented:
> I've updated pull request.
> Added option to filter values or see only selected one, and changed format to CSV.
> TXBuffer.startJsonStream(); need to stay as it is, because without it browser won't recognize newline characters.

CSV format is different.
CSV only outputs the labels on the first line (separated using the same char as on the value lines)
All next lines only contain the values in columns, separated by the separator char.

The label line is optional.

Grovkillen commented:
Yes that isn't CSV standard. This is more a key-value standard. Commonly
used in "databases".



Den lör 21 mars 2020 14:02TD-er <notifications@github.com> skrev:

> I've updated pull request.
> Added option to filter values or see only selected one, and changed format
> to CSV.
> TXBuffer.startJsonStream(); need to stay as it is, because without it
> browser won't recognize newline characters.
>
> CSV format is different.
> CSV only outputs the labels on the first line (separated using the same
> char as on the value lines)
> All next lines only contain the values in columns, separated by the
> separator char.
>
> The label line is optional.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letscontrolit/ESPEasy/pull/2956#issuecomment-602040853>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLYCYXB3GT7H5PEUJQEP7LRIS3FTANCNFSM4LQVGSOA>
> .
>

mobrembski commented:
> > I've updated pull request.
> > Added option to filter values or see only selected one, and changed format to CSV.
> > TXBuffer.startJsonStream(); need to stay as it is, because without it browser won't recognize newline characters.
> 
> CSV format is different.
> CSV only outputs the labels on the first line (separated using the same char as on the value lines)
> All next lines only contain the values in columns, separated by the separator char.
> 
> The label line is optional.

Ahh right, ok, i will fix that :)

TD-er commented:
Maybe also add a flag then to enable/disable the CSV header line, so you can request it when starting to receive data and just get the values in subsequent requests.

If it is changing to CSV, better name it so in the URL instead of RAW.

mobrembski committed:
Fixed invalid check of valnr validity in CSV output mode

mobrembski committed:
Renamed RAW to CSV, added printing of header

mobrembski commented:
Updated again ;)
Renamed from RAW to CSV, added printing of header (which could be disabled, default is enabled)

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,75 @@
+// ********************************************************************************
+void handle_csvval()
+{
+  String htmlData(25, '\0'); // Reserve for error message
```
I'm not sure what this constructor should do, but it looks like you're calling it to store the integer value 25 in base-0 notation (no idea what that should do).
It is not a reserve + init to an empty string.

mobrembski replied:
I'm sorry i have a problems with my VSCode with detecting using headers.
Is it not a std::string?
Because if so, then it should create a string with 25 chars which of these are '\0' (string terminator)

TD-er replied:
Nope, `String` is an Arduino specific object.
We don't use std::string in our code here. Not that it is forbidden, but more that this may be confusing when using `String` and `string` in the same project.

mobrembski replied:
Oh, ok :)
I'm pretty fresh in ESPEasy, so i haven't know. but ok i fixed it ;)

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,75 @@
+        htmlData += "Name";
+        htmlData += ';';
+        htmlData += "Value";
+        htmlData += '\n';
```
Every place where you add a `'\n'` is a good place to call `addHtml` and empty the string.
The general idea is to send the data to the connected client in chunks so we don't have to allocate a lot of memory.
It also makes the ESP very responsive as we generate the output at the same time the client downloads it.

mobrembski replied:
Ok, very good idea,  i will fix that :)

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,75 @@
+          htmlData += String(ExtraTaskSettings.TaskDeviceValueNames[x]);
+          htmlData += ';';
+          htmlData += formatUserVarNoCheck(taskNr, x);
+          htmlData += '\n';
```
You still output the data in this form:

```
Name;Value
var1;1.234
var2;2.345
var3;3.456
```

CSV is more like this:

```
var1;var1;var3
1.234;2.345;3.456
```

This way you only need to send the 'header' once, or maybe even leave it out on a (next) request.

mobrembski replied:
Ahh, so:
Header(optional) -> var1;var1;var3
values                  ->1.234;2.345;3.456

Right?

uzi18 replied:
yes, but fought it is raw =)
maybe format it like:
`name=value`

TD-er replied:
> yes, but fought it is raw =)
> maybe format it like:
> `name=value`

He just changed all into making it look like CSV (even the URL endpoint) :)

mobrembski replied:
Ok, Fixed ;)

mobrembski committed:
Removed usage of String.clear()

It is not available in minimal_core_242_ESP8266_1M_OTA env

mobrembski commented:
Hi all,

Can we merge this PR, please? ;)

TD-er commented:
> Hi all,
> 
> Can we merge this PR, please? ;)

Sure
